Samples/Widgets/cpp-win32-packaged/SampleWidgetProviderApp/WidgetProvider.cpp

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "WidgetProvider.h"
+
+using CreateWidgetImplFn = std::function<std::shared_ptr<WidgetImplBase>(winrt::WidgetContext, winrt::hstring)>;
+
+struct WidgetImplCreationInfo
+{
+    const wchar_t* const widgetName;
+    CreateWidgetImplFn factoryFn;
+};
+
+template<typename T>
+std::shared_ptr<WidgetImplBase> CreateWidgetFn(winrt::WidgetContext widgetContext, winrt::hstring state)
+{
+    auto widgetId = widgetContext.Id();
+    auto newWidget = std::make_shared<T>(widgetId, state);
+
+    // Call UpdateWidget() to send data/template to the newly created widget.
+    winrt::WidgetUpdateRequestOptions updateOptions{ widgetId };
+    updateOptions.Template(newWidget->GetTemplateForWidget());
+    updateOptions.Data(newWidget->GetDataForWidget());
+    // You can store some custom state in the widget service that you will be able to query at any time.
+    updateOptions.CustomState(newWidget->State());
+    // Update the widget
+    winrt::WidgetManager::GetDefault().UpdateWidget(updateOptions);
+    return newWidget;
+}
+
+#include "CountingWidgetImpl.h"
+#include "WeatherWidgetImpl.h"
+
+// Register all widget types here, optionally specify an is enabled function
+const static std::array<WidgetImplCreationInfo, 2> s_widgetImplRegistry = { { {L"Counting_Widget", &CreateWidgetFn<CountingWidget>}, {L"Weather_Widget", &CreateWidgetFn<WeatherWidget>} } };
+
+std::unordered_map<winrt::hstring, std::shared_ptr<WidgetImplBase>> WidgetProvider::m_runningWidgetImpl{};
+
+WidgetProvider::WidgetProvider()
+{
+    RecoverRunningWidgets();
+}
+
+std::shared_ptr<WidgetImplBase> WidgetProvider::InitializeWidgetInternal(winrt::Microsoft::Windows::Widgets::Providers::WidgetContext widgetContext, winrt::hstring state)
+{
+    auto widgetName = widgetContext.DefinitionId();
+    auto widgetId = widgetContext.Id();
+    auto lookUpFunction = [widgetName](WidgetImplCreationInfo widgetCreationInfo) { return widgetName == widgetCreationInfo.widgetName; };
+    auto creationInfoIt = std::find_if(s_widgetImplRegistry.begin(), s_widgetImplRegistry.end(), lookUpFunction);
```

Comment:
nit, can be simplified with a lambda:

```suggestion
    auto creationInfoIt = std::find_if(s_widgetImplRegistry.begin(), s_widgetImplRegistry.end(), [widgetName](const auto&& e) { return widgetName == e.widgetName; });
```